src/backend/utils/time/sharedqueryplan.c

```diff
@@ -0,0 +1,203 @@
+			                    "but it is currently %d. Already waited for %ldus and will time out at %ldus.",
+			                    gp_session_id,
+			                    gp_command_count,
+			                    entry ? entry->commandCount : -1,
+			                    total_sleep_time_us,
+			                    timeout_us));
+			warning_sleep_time_us = 0;
+		}
+		LWLockRelease(SharedQueryPlanLock);
+
+		if (total_sleep_time_us >= timeout_us)
+			ereport(ERROR, errcode(ERRCODE_GP_INTERCONNECTION_ERROR),
+			        errmsg("ReadSharedQueryPlan timed out after %ldus waiting for writer QE to set shared plan.", total_sleep_time_us),
+			        errdetail("We are waiting for writer QE of session %d to set shared "
+			                  "serializedPlantree and serializedQueryDispatchDesc for commandCount %d "
+			                  "but it is currently %d. This may due to network congestion, please retry.",
+			                  gp_session_id,
+			                  gp_command_count,
+			                  entry ? entry->commandCount : -1));
+
+		pg_usleep(sleep_per_check_us);
+
+		CHECK_FOR_INTERRUPTS();
+
+		total_sleep_time_us += sleep_per_check_us;
+		warning_sleep_time_us += sleep_per_check_us;
+	}
+}
+
+void
+WriteSharedQueryPlan(const char *serializedPlantree, int serializedPlantreelen, const char *serializedQueryDispatchDesc, int serializedQueryDispatchDesclen, const SliceTable *sliceTable)
+{
+	int i;
+	ExecSlice *slice = NULL;
+	ExecSlice *local = NULL;
+
+	LWLockAcquire(SharedQueryPlanLock, LW_EXCLUSIVE);
+	local = &sliceTable->slices[sliceTable->localSlice];
+	int referenced = local->parallel_workers - 1; // referenced counts how many *other* QE need to read from this SharedQueryPlan
+	for (i = 0; i < sliceTable->numSlices; i++)
+	{
+		slice = &sliceTable->slices[i];
+		if (slice->sliceIndex == sliceTable->localSlice)
+			continue;
+		// there can be multiple roots when plan contains InitPlan, only consider those who're in the same root
+		if (slice->rootIndex != local->rootIndex)
+			continue;
+		if (slice->gangType == GANGTYPE_PRIMARY_WRITER || slice->gangType == GANGTYPE_PRIMARY_READER || slice->gangType == GANGTYPE_SINGLETON_READER)
+		{
+			ListCell *segment_cell;
+			foreach(segment_cell, slice->segments)
+			{
+				referenced += lfirst_int(segment_cell) == GpIdentity.segindex;
+			}
+		}
+	}
+	Assert(referenced >= 0);
+
+	if (referenced > 0)
+	{
+		bool found;
+		SharedQueryPlanEntry *entry;
+		ResourceOwner oldowner;
+		Size size = serializedPlantreelen + serializedQueryDispatchDesclen;
+
+		entry = (SharedQueryPlanEntry *) hash_search(SharedQueryPlanHash,
+		                                             &gp_session_id,
+		                                             HASH_ENTER,
+		                                             &found);
+		Assert(!found);
+		if (!SharedQueryPlanResOwner)
+			SharedQueryPlanResOwner = ResourceOwnerCreate(NULL, "SharedQueryPlanResOwner");
```

Comment:
entry add a new var to store `ResourceOwner` , that's better ?

Reply:
the parent of `SharedQueryPlanResOwner` so it has a unlimited lifetime. this is to make `PrintDSMLeakWarning` **not** printing any warnings at transaction completion because the DSM's lifetime is attached to `entry->reference` that we already have a reclaim logic for. 

shared snapshot have the same approach to create a unbounded `DumpResOwner` which I guess is to do the same thing.

Reply:
It requires another round of `hash_search` to find the entry and decrease `referenced` _after_ the deserialization of `serializedPlantree` and `serializedQueryDesc`. I think maybe two `memcpy` is still more efficient.